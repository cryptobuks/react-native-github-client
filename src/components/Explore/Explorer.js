import React, { PureComponent } from 'react';
import { StyleSheet, FlatList, TouchableOpacity } from 'react-native';
import { List, Text, ListItem, Left, Body, Thumbnail } from 'native-base';
import moment from 'moment';
import Container from 'SafeContainer';
import { AsyncStorage } from 'react-native';
import { getEvent } from './queries';
import { addLogoutListener, removeLogoutListener } from '../../auth';
import { openWebView } from 'utils';

type Props = {
  navigator: Object,
};

/**
 * An explorer to navigate other github content
 */
class Explorer extends PureComponent<Props> {
  page = 1;

  state = {
    events: [],
    refreshing: false,
  };
  componentDidMount = () => {
    AsyncStorage.getItem('events')
      .then(events => this.setState({ events: JSON.parse(events) || [] }))
      .then(() => this.fetchEvent(true))
      .catch(err => console.log(err));
    this.props.navigator.setOnNavigatorEvent(this.onNavigatorEvent);
    addLogoutListener('refreshExplore', this.handleRefresh);
  };

  componentWillUnmount = () => {
    removeLogoutListener('refreshExplore');
  };

  handlePressUser = (login: String) => () =>
    this.props.navigator.push({
      screen: 'profile.user',
      title: login,
      passProps: { login },
    });

  onNavigatorEvent = ({ type, id }) => {
    if (type === 'NavBarButtonPress' && id === 'search') {
      this.props.navigator.push({
        screen: 'explore.search',
        title: 'Search',
      });
    }
  };

  handleRefresh = () => this.fetchEvent(true);
  handleLoadMore = () => this.fetchEvent(false);

  fetchEvent = (refresh = false) => {
    if (this.loading || this.state.refreshing) return;
    this.loading = true;

    const page = refresh ? 1 : this.page;
    if (refresh) this.setState({ refreshing: true });

    getEvent(page)
      .then(res => {
        this.loading = false;
        this.page = page + 1;
        this.setState(
          {
            events: (refresh ? [] : this.state.events).concat(res.data),
            refreshing: false,
          },
          this.storeEvents
        );
      })
      .catch(err => {
        this.loading = false;
        this.setState({ error: err.message, refreshing: false });
      });
  };

  storeEvents = () =>
    AsyncStorage.setItem('events', JSON.stringify(this.state.events)).catch(
      err => console.log(err)
    );

  renderEvent = ({ item }) => {
    const avatar = (
      <TouchableOpacity onPress={this.handlePressUser(item.actor.login)}>
        <Thumbnail
          small
          source={{ uri: item.actor.avatar_url }}
          style={styles.avatar}
        />
      </TouchableOpacity>
    );

    switch (item.type) {
      case 'WatchEvent':
        return (
          <ListItem
            avatar
            style={styles.listitem}
            onPress={openWebView(
              item.repo.url.replace('api.github.com/repos', 'github.com'),
              this.props.navigator
            )}
          >
            <Left>{avatar}</Left>
            <Body>
              <Text>
                {item.actor.login} starred {item.repo.name}
              </Text>
              <Text note>{moment(item.created_at).fromNow()}</Text>
            </Body>
          </ListItem>
        );
      case 'ForkEvent':
        return (
          <ListItem
            avatar
            style={styles.listitem}
            onPress={openWebView(
              item.payload.forkee.url.replace(
                'api.github.com/repos',
                'github.com'
              ),
              this.props.navigator
            )}
          >
            <Left>{avatar}</Left>
            <Body>
              <Text>
                {item.actor.login} forked {item.payload.forkee.name} from{' '}
                {item.repo.name}
              </Text>
              <Text note>{moment(item.created_at).fromNow()}</Text>
            </Body>
          </ListItem>
        );
      case 'CreateEvent':
        return (
          <ListItem
            avatar
            style={styles.listitem}
            onPress={openWebView(
              item.repo.url.replace('api.github.com/repos', 'github.com'),
              this.props.navigator
            )}
          >
            <Left>{avatar}</Left>
            <Body>
              <Text>
                {item.actor.login} created a repository {item.repo.name}
              </Text>
              <Text note>{moment(item.created_at).fromNow()}</Text>
            </Body>
          </ListItem>
        );
      case 'PushEvent':
        return (
          <ListItem
            avatar
            style={styles.listitem}
            onPress={openWebView(
              item.repo.url.replace('api.github.com/repos', 'github.com') +
                '/commit/' +
                item.payload.head,
              this.props.navigator
            )}
          >
            <Left>{avatar}</Left>
            <Body>
              <Text>
                {item.actor.login} pushed to {item.repo.name}
              </Text>
              <Text note>{moment(item.created_at).fromNow()}</Text>
            </Body>
          </ListItem>
        );
    }
  };

  eventKeyExtractor = event => event.id;

  render = () => {
    const { error, events } = this.state;

    return events.length === 0 ? (
      error ? (
        <Text>{error}</Text>
      ) : (
        <Text>Loading...</Text>
      )
    ) : (
      <Container>
        <List>
          <FlatList
            data={this.state.events}
            keyExtractor={this.eventKeyExtractor}
            renderItem={this.renderEvent}
            onEndReached={this.handleLoadMore}
            onRefresh={this.handleRefresh}
            refreshing={this.state.refreshing}
          />
        </List>
      </Container>
    );
  };
}

const styles = StyleSheet.create({
  avatar: {
    margin: 5,
  },
  listitem: {
    paddingVertical: 5,
  },
});

export default Explorer;
