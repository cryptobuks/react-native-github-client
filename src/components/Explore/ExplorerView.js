import React, { PureComponent } from 'react';
import { StyleSheet, FlatList, TouchableOpacity } from 'react-native';
import { List, Text, ListItem, Left, Body, Thumbnail } from 'native-base';
import moment from 'moment';
import Container from 'SafeContainer';
import { openWebView } from 'utils';

type Props = {
  navigator: Object,
  events: [Object],
  onLoadMore: null => null,
  onRefresh: null => null,
  refreshing: Boolean,
};

/**
 * An explorer to navigate other github content
 */
class ExplorerView extends PureComponent<Props> {
  handlePressUser = (login: String) => () =>
    this.props.navigator.push({
      screen: 'profile.user',
      title: login,
      passProps: { login },
    });

  renderEvent = ({ item }) => {
    const avatar = (
      <TouchableOpacity onPress={this.handlePressUser(item.actor.login)}>
        <Thumbnail
          small
          source={{ uri: item.actor.avatar_url }}
          style={styles.avatar}
        />
      </TouchableOpacity>
    );

    switch (item.type) {
      case 'WatchEvent':
        return (
          <ListItem
            avatar
            style={styles.listitem}
            onPress={openWebView(
              item.repo.url.replace('api.github.com/repos', 'github.com'),
              this.props.navigator
            )}
          >
            <Left>{avatar}</Left>
            <Body>
              <Text>
                {item.actor.login} starred {item.repo.name}
              </Text>
              <Text note>{moment(item.created_at).fromNow()}</Text>
            </Body>
          </ListItem>
        );
      case 'ForkEvent':
        return (
          <ListItem
            avatar
            style={styles.listitem}
            onPress={openWebView(
              item.payload.forkee.url.replace(
                'api.github.com/repos',
                'github.com'
              ),
              this.props.navigator
            )}
          >
            <Left>{avatar}</Left>
            <Body>
              <Text>
                {item.actor.login} forked {item.payload.forkee.name} from{' '}
                {item.repo.name}
              </Text>
              <Text note>{moment(item.created_at).fromNow()}</Text>
            </Body>
          </ListItem>
        );
      case 'CreateEvent':
        return (
          <ListItem
            avatar
            style={styles.listitem}
            onPress={openWebView(
              item.repo.url.replace('api.github.com/repos', 'github.com'),
              this.props.navigator
            )}
          >
            <Left>{avatar}</Left>
            <Body>
              <Text>
                {item.actor.login} created a repository {item.repo.name}
              </Text>
              <Text note>{moment(item.created_at).fromNow()}</Text>
            </Body>
          </ListItem>
        );
      case 'PushEvent':
        return (
          <ListItem
            avatar
            style={styles.listitem}
            onPress={openWebView(
              item.repo.url.replace('api.github.com/repos', 'github.com') +
                '/commit/' +
                item.payload.head,
              this.props.navigator
            )}
          >
            <Left>{avatar}</Left>
            <Body>
              <Text>
                {item.actor.login} pushed to {item.repo.name}
              </Text>
              <Text note>{moment(item.created_at).fromNow()}</Text>
            </Body>
          </ListItem>
        );
    }
  };

  eventKeyExtractor = event => event.id;

  render = () => (
    <Container>
      <List>
        <FlatList
          data={this.props.events}
          keyExtractor={this.eventKeyExtractor}
          renderItem={this.renderEvent}
          onEndReached={this.props.onLoadMore}
          onRefresh={this.props.onRefresh}
          refreshing={this.props.refreshing}
        />
      </List>
    </Container>
  );
}

const styles = StyleSheet.create({
  avatar: {
    margin: 5,
  },
  listitem: {
    paddingVertical: 5,
  },
});

export default ExplorerView;
