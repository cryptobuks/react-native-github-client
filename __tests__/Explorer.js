import 'react-native';
import React from 'react';
import Explorer from '../src/components/Explore/Explorer';
import ExplorerView from '../src/components/Explore/ExplorerView';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

const navigator = {
  setOnNavigatorEvent: jest.fn(),
};

it('explorer renders correctly', () => {
  expect(
    renderer.create(<Explorer navigator={navigator} />).toJSON()
  ).toMatchSnapshot();
});

it('explorer list display events', () => {
  const events = [
    {
      id: '123456',
      type: 'CreateEvent',
      actor: {
        login: 'foo',
        avatar_url: '',
      },
      repo: {
        name: 'foo/bar',
        url: '',
      },
      created_at: '2018-03-26T04:06:30Z',
    },
    {
      id: '2345678',
      type: 'WatchEvent',
      actor: {
        login: 'segmentation',
        avatar_url: '',
      },
      repo: {
        name: 'asdf',
        url: '',
      },
      created_at: '2018-03-26T02:48:28Z',
    },
    {
      id: '12987423',
      type: 'ForkEvent',
      actor: {
        login: 'foo',
        avatar_url: '',
      },
      payload: {
        forkee: {
          url: '',
          name: 'asdf',
        },
      },
      repo: {
        name: 'asdf',
      },
      created_at: '2018-03-26T02:48:28Z',
    },
    {
      id: '230405',
      type: 'PushEvent',
      actor: {
        login: 'foo',
        avatar_url: '',
      },
      payload: {
        head: '1a2b3c',
      },
      repo: {
        name: 'asdf',
        url: '',
      },
      created_at: '2018-03-26T02:48:28Z',
    },
  ];

  expect(
    renderer
      .create(<ExplorerView navigator={navigator} events={events} />)
      .toJSON()
  ).toMatchSnapshot();
});
