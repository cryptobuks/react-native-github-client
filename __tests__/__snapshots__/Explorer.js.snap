// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`explorer list display events 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "height": 1334,
    }
  }
  virtual={undefined}
>
  <RCTSafeAreaView
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      closeOnRowBeginSwipe={false}
      closeOnRowPress={true}
      closeOnScroll={true}
      directionalDistanceChangeThreshold={2}
      disableLeftSwipe={false}
      disableRightSwipe={false}
      leftOpenValue={0}
      previewFirstRow={false}
      recalculateHiddenLayout={false}
      rightOpenValue={0}
      style={
        Array [
          Object {},
          Object {
            "flex": 1,
          },
        ]
      }
      swipeToOpenPercent={50}
      virtual={undefined}
    >
      <RCTScrollView
        ListEmptyComponent={
          <Styled(Text)
            virtual={undefined}
          >
            No Content
          </Styled(Text)>
        }
        data={
          Array [
            Object {
              "actor": Object {
                "avatar_url": "",
                "login": "foo",
              },
              "created_at": "2018-03-26T04:06:30Z",
              "id": "123456",
              "repo": Object {
                "name": "foo/bar",
                "url": "",
              },
              "type": "CreateEvent",
            },
            Object {
              "actor": Object {
                "avatar_url": "",
                "login": "segmentation",
              },
              "created_at": "2018-03-26T02:48:28Z",
              "id": "2345678",
              "repo": Object {
                "name": "asdf",
                "url": "",
              },
              "type": "WatchEvent",
            },
            Object {
              "actor": Object {
                "avatar_url": "",
                "login": "foo",
              },
              "created_at": "2018-03-26T02:48:28Z",
              "id": "12987423",
              "payload": Object {
                "forkee": Object {
                  "name": "asdf",
                  "url": "",
                },
              },
              "repo": Object {
                "name": "asdf",
              },
              "type": "ForkEvent",
            },
            Object {
              "actor": Object {
                "avatar_url": "",
                "login": "foo",
              },
              "created_at": "2018-03-26T02:48:28Z",
              "id": "230405",
              "payload": Object {
                "head": "1a2b3c",
              },
              "repo": Object {
                "name": "asdf",
                "url": "",
              },
              "type": "PushEvent",
            },
          ]
        }
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        invertStickyHeaders={undefined}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        numColumns={1}
        onContentSizeChange={[Function]}
        onEndReached={undefined}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onRefresh={undefined}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        refreshing={undefined}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={null}
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View
                avatar={true}
                onPress={[Function]}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "borderBottomWidth": null,
                      "borderColor": "#c9c9c9",
                      "flexDirection": "row",
                      "marginLeft": 15,
                      "paddingBottom": null,
                      "paddingRight": null,
                      "paddingTop": null,
                    },
                    Object {
                      "paddingVertical": 5,
                    },
                  ]
                }
                virtual={undefined}
              >
                <View
                  style={
                    Object {
                      "alignItems": "flex-start",
                      "alignSelf": "center",
                      "flex": 0,
                      "flexDirection": "row",
                    }
                  }
                  virtual={undefined}
                >
                  <View
                    accessibilityComponentType={undefined}
                    accessibilityLabel={undefined}
                    accessibilityTraits={undefined}
                    accessible={true}
                    collapsable={undefined}
                    hasTVPreferredFocus={undefined}
                    hitSlop={undefined}
                    isTVSelectable={true}
                    nativeID={undefined}
                    onLayout={undefined}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                    testID={undefined}
                    tvParallaxProperties={undefined}
                  >
                    <Image
                      small={true}
                      source={
                        Object {
                          "uri": "",
                        }
                      }
                      style={
                        Array [
                          Object {
                            "borderRadius": 18,
                            "height": 36,
                            "width": 36,
                          },
                          Object {
                            "margin": 5,
                          },
                        ]
                      }
                      virtual={undefined}
                    />
                  </View>
                </View>
                <View
                  style={
                    Object {
                      "alignItems": null,
                      "alignSelf": null,
                      "borderBottomWidth": 0.5,
                      "borderColor": "#c9c9c9",
                      "flex": 1,
                      "marginLeft": 15,
                      "paddingBottom": 10,
                      "paddingTop": 10,
                    }
                  }
                  virtual={undefined}
                >
                  <Text
                    accessible={true}
                    allowFontScaling={true}
                    ellipsizeMode="tail"
                    style={
                      Object {
                        "color": "#000",
                        "fontFamily": "System",
                        "fontSize": 16,
                        "marginLeft": null,
                        "marginRight": 10,
                      }
                    }
                    uppercase={false}
                    virtual={undefined}
                  >
                    foo
                     created a repository 
                    foo/bar
                  </Text>
                  <Text
                    accessible={true}
                    allowFontScaling={true}
                    ellipsizeMode="tail"
                    note={true}
                    style={
                      Object {
                        "color": "#808080",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "200",
                        "marginLeft": null,
                        "marginRight": 10,
                      }
                    }
                    uppercase={false}
                    virtual={undefined}
                  >
                    in a year
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={null}
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View
                avatar={true}
                onPress={[Function]}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "borderBottomWidth": null,
                      "borderColor": "#c9c9c9",
                      "flexDirection": "row",
                      "marginLeft": 15,
                      "paddingBottom": null,
                      "paddingRight": null,
                      "paddingTop": null,
                    },
                    Object {
                      "paddingVertical": 5,
                    },
                  ]
                }
                virtual={undefined}
              >
                <View
                  style={
                    Object {
                      "alignItems": "flex-start",
                      "alignSelf": "center",
                      "flex": 0,
                      "flexDirection": "row",
                    }
                  }
                  virtual={undefined}
                >
                  <View
                    accessibilityComponentType={undefined}
                    accessibilityLabel={undefined}
                    accessibilityTraits={undefined}
                    accessible={true}
                    collapsable={undefined}
                    hasTVPreferredFocus={undefined}
                    hitSlop={undefined}
                    isTVSelectable={true}
                    nativeID={undefined}
                    onLayout={undefined}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                    testID={undefined}
                    tvParallaxProperties={undefined}
                  >
                    <Image
                      small={true}
                      source={
                        Object {
                          "uri": "",
                        }
                      }
                      style={
                        Array [
                          Object {
                            "borderRadius": 18,
                            "height": 36,
                            "width": 36,
                          },
                          Object {
                            "margin": 5,
                          },
                        ]
                      }
                      virtual={undefined}
                    />
                  </View>
                </View>
                <View
                  style={
                    Object {
                      "alignItems": null,
                      "alignSelf": null,
                      "borderBottomWidth": 0.5,
                      "borderColor": "#c9c9c9",
                      "flex": 1,
                      "marginLeft": 15,
                      "paddingBottom": 10,
                      "paddingTop": 10,
                    }
                  }
                  virtual={undefined}
                >
                  <Text
                    accessible={true}
                    allowFontScaling={true}
                    ellipsizeMode="tail"
                    style={
                      Object {
                        "color": "#000",
                        "fontFamily": "System",
                        "fontSize": 16,
                        "marginLeft": null,
                        "marginRight": 10,
                      }
                    }
                    uppercase={false}
                    virtual={undefined}
                  >
                    segmentation
                     starred 
                    asdf
                  </Text>
                  <Text
                    accessible={true}
                    allowFontScaling={true}
                    ellipsizeMode="tail"
                    note={true}
                    style={
                      Object {
                        "color": "#808080",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "200",
                        "marginLeft": null,
                        "marginRight": 10,
                      }
                    }
                    uppercase={false}
                    virtual={undefined}
                  >
                    in a year
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={null}
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View
                avatar={true}
                onPress={[Function]}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "borderBottomWidth": null,
                      "borderColor": "#c9c9c9",
                      "flexDirection": "row",
                      "marginLeft": 15,
                      "paddingBottom": null,
                      "paddingRight": null,
                      "paddingTop": null,
                    },
                    Object {
                      "paddingVertical": 5,
                    },
                  ]
                }
                virtual={undefined}
              >
                <View
                  style={
                    Object {
                      "alignItems": "flex-start",
                      "alignSelf": "center",
                      "flex": 0,
                      "flexDirection": "row",
                    }
                  }
                  virtual={undefined}
                >
                  <View
                    accessibilityComponentType={undefined}
                    accessibilityLabel={undefined}
                    accessibilityTraits={undefined}
                    accessible={true}
                    collapsable={undefined}
                    hasTVPreferredFocus={undefined}
                    hitSlop={undefined}
                    isTVSelectable={true}
                    nativeID={undefined}
                    onLayout={undefined}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                    testID={undefined}
                    tvParallaxProperties={undefined}
                  >
                    <Image
                      small={true}
                      source={
                        Object {
                          "uri": "",
                        }
                      }
                      style={
                        Array [
                          Object {
                            "borderRadius": 18,
                            "height": 36,
                            "width": 36,
                          },
                          Object {
                            "margin": 5,
                          },
                        ]
                      }
                      virtual={undefined}
                    />
                  </View>
                </View>
                <View
                  style={
                    Object {
                      "alignItems": null,
                      "alignSelf": null,
                      "borderBottomWidth": 0.5,
                      "borderColor": "#c9c9c9",
                      "flex": 1,
                      "marginLeft": 15,
                      "paddingBottom": 10,
                      "paddingTop": 10,
                    }
                  }
                  virtual={undefined}
                >
                  <Text
                    accessible={true}
                    allowFontScaling={true}
                    ellipsizeMode="tail"
                    style={
                      Object {
                        "color": "#000",
                        "fontFamily": "System",
                        "fontSize": 16,
                        "marginLeft": null,
                        "marginRight": 10,
                      }
                    }
                    uppercase={false}
                    virtual={undefined}
                  >
                    foo
                     forked 
                    asdf
                     from
                     
                    asdf
                  </Text>
                  <Text
                    accessible={true}
                    allowFontScaling={true}
                    ellipsizeMode="tail"
                    note={true}
                    style={
                      Object {
                        "color": "#808080",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "200",
                        "marginLeft": null,
                        "marginRight": 10,
                      }
                    }
                    uppercase={false}
                    virtual={undefined}
                  >
                    in a year
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={null}
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View
                avatar={true}
                onPress={[Function]}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "borderBottomWidth": null,
                      "borderColor": "#c9c9c9",
                      "flexDirection": "row",
                      "marginLeft": 15,
                      "paddingBottom": null,
                      "paddingRight": null,
                      "paddingTop": null,
                    },
                    Object {
                      "paddingVertical": 5,
                    },
                  ]
                }
                virtual={undefined}
              >
                <View
                  style={
                    Object {
                      "alignItems": "flex-start",
                      "alignSelf": "center",
                      "flex": 0,
                      "flexDirection": "row",
                    }
                  }
                  virtual={undefined}
                >
                  <View
                    accessibilityComponentType={undefined}
                    accessibilityLabel={undefined}
                    accessibilityTraits={undefined}
                    accessible={true}
                    collapsable={undefined}
                    hasTVPreferredFocus={undefined}
                    hitSlop={undefined}
                    isTVSelectable={true}
                    nativeID={undefined}
                    onLayout={undefined}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                    testID={undefined}
                    tvParallaxProperties={undefined}
                  >
                    <Image
                      small={true}
                      source={
                        Object {
                          "uri": "",
                        }
                      }
                      style={
                        Array [
                          Object {
                            "borderRadius": 18,
                            "height": 36,
                            "width": 36,
                          },
                          Object {
                            "margin": 5,
                          },
                        ]
                      }
                      virtual={undefined}
                    />
                  </View>
                </View>
                <View
                  style={
                    Object {
                      "alignItems": null,
                      "alignSelf": null,
                      "borderBottomWidth": 0.5,
                      "borderColor": "#c9c9c9",
                      "flex": 1,
                      "marginLeft": 15,
                      "paddingBottom": 10,
                      "paddingTop": 10,
                    }
                  }
                  virtual={undefined}
                >
                  <Text
                    accessible={true}
                    allowFontScaling={true}
                    ellipsizeMode="tail"
                    style={
                      Object {
                        "color": "#000",
                        "fontFamily": "System",
                        "fontSize": 16,
                        "marginLeft": null,
                        "marginRight": 10,
                      }
                    }
                    uppercase={false}
                    virtual={undefined}
                  >
                    foo
                     pushed to 
                    asdf
                  </Text>
                  <Text
                    accessible={true}
                    allowFontScaling={true}
                    ellipsizeMode="tail"
                    note={true}
                    style={
                      Object {
                        "color": "#808080",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "200",
                        "marginLeft": null,
                        "marginRight": 10,
                      }
                    }
                    uppercase={false}
                    virtual={undefined}
                  >
                    in a year
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </RCTSafeAreaView>
</View>
`;

exports[`explorer renders correctly 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "#000",
      "fontFamily": "System",
      "fontSize": 16,
    }
  }
  uppercase={false}
  virtual={undefined}
>
  Loading...
</Text>
`;
